File: services/flutter-dev-container/code/packages/couchbase_lite_p2p/ios/Classes/CouchbaseLiteP2pPlugin.swift
import Flutter
import UIKit
import CouchbaseLiteSwift

public class CouchbaseLiteP2pPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var database: Database?
    private var multipeerSyncManager: MultipeerSyncManager?
    private var replicator: Replicator?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "couchbase_lite_p2p", binaryMessenger: registrar.messenger())
        let instance = CouchbaseLiteP2pPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.initializeCouchbase()
    }
    
    private func initializeCouchbase() {
        do {
            // Open Database
            let config = DatabaseConfiguration()
            database = try Database(name: "counter_db", config: config)
            
            // Initialize Sync Manager
            if let db = database {
                multipeerSyncManager = MultipeerSyncManager(database: db)
                
                // Listen for DB changes
                setupDatabaseListeners()
                
                // Listen for P2P status changes
                setupP2PListeners()
            }
        } catch {
            print("Error initializing Couchbase Lite: \(error)")
        }
    }
    
    private func setupDatabaseListeners() {
        guard let db = database else { return }
        
        do {
            let collection = try db.defaultCollection()
            _ = collection.addDocumentChangeListener(id: "counter") { [weak self] change in
                guard let self = self else { return }
                do {
                    if let doc = try collection.document(id: change.documentID) {
                        let count = doc.int(forKey: "count")
                        DispatchQueue.main.async {
                            self.channel?.invokeMethod("onCountChanged", arguments: count)
                        }
                    }
                } catch {
                    print("Error getting document: \(error)")
                }
            }
        } catch {
            print("Error setting up listener: \(error)")
        }
    }
    
    private func setupP2PListeners() {
        // TODO: Implement P2P status listener from MultipeerSyncManager
        // For now, we'll just poll or rely on manual updates if needed, 
        // but ideally MultipeerSyncManager should expose a callback/delegate.
        
        multipeerSyncManager?.onStatusChanged = { [weak self] status in
            DispatchQueue.main.async {
                self?.channel?.invokeMethod("onP2PStatusChanged", arguments: status)
            }
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "increment":
            incrementCounter(result: result)
        case "getCount":
            getCounter(result: result)
        case "startP2P":
            multipeerSyncManager?.start()
            result(true)
        case "getP2PStatus":
            if let status = multipeerSyncManager?.currentStatus {
                result(status)
            } else {
                result(["isRunning": false, "status": "Stopped"])
            }
        case "startSyncGatewayReplication":
            if let args = call.arguments as? [String: Any],
               let urlString = args["url"] as? String {
                startSyncGatewayReplication(urlString: urlString, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: "URL is required", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startSyncGatewayReplication(urlString: String, result: @escaping FlutterResult) {
        guard let db = database else {
            result(FlutterError(code: "DB_ERROR", message: "Database not initialized", details: nil))
            return
        }
        
        guard let url = URL(string: urlString) else {
            result(FlutterError(code: "INVALID_URL", message: "Invalid URL: \(urlString)", details: nil))
            return
        }
        
        do {
            let target = URLEndpoint(url: url)
            var config = ReplicatorConfiguration(target: target)
            config.replicatorType = .pushAndPull
            config.continuous = true
            
            // Add default collection
            let collection = try db.defaultCollection()
            config.addCollection(collection, config: nil)
            
            replicator = Replicator(config: config)
            
            replicator?.addChangeListener { [weak self] change in
                guard let self = self else { return }
                let status = change.status
                var statusStr = "Stopped"
                switch status.activity {
                case .stopped: statusStr = "Stopped"
                case .offline: statusStr = "Offline"
                case .connecting: statusStr = "Connecting"
                case .idle: statusStr = "Idle"
                case .busy: statusStr = "Busy"
                @unknown default: statusStr = "Unknown"
                }
                
                var errorStr: String? = nil
                if let error = status.error {
                    errorStr = error.localizedDescription
                }
                
                DispatchQueue.main.async {
                    self.channel?.invokeMethod("onSyncGatewayStatusChanged", arguments: [
                        "status": statusStr,
                        "error": errorStr
                    ])
                }
            }
            
            replicator?.start()
            result(true)
        } catch {
            result(FlutterError(code: "REPLICATION_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func incrementCounter(result: @escaping FlutterResult) {
        guard let db = database else {
            result(FlutterError(code: "DB_ERROR", message: "Database not initialized", details: nil))
            return
        }
        
        do {
            let collection = try db.defaultCollection()
            var newCount = 0
            
            try db.inBatch {
                var doc = try collection.document(id: "counter")?.toMutable()
                if doc == nil {
                    doc = MutableDocument(id: "counter")
                }
                
                if let doc = doc {
                    let count = doc.int(forKey: "count")
                    newCount = count + 1
                    doc.setInt(newCount, forKey: "count")
                    try collection.save(document: doc)
                }
            }
            result(newCount)
        } catch {
            result(FlutterError(code: "DB_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func getCounter(result: @escaping FlutterResult) {
        guard let db = database else {
            result(0)
            return
        }
        
        do {
            let collection = try db.defaultCollection()
            if let doc = try collection.document(id: "counter") {
                result(doc.int(forKey: "count"))
            } else {
                result(0)
            }
        } catch {
            result(FlutterError(code: "DB_ERROR", message: error.localizedDescription, details: nil))
        }
    }
}

File: services/flutter-dev-container/code/packages/couchbase_lite_p2p/android/src/main/kotlin/com/example/couchbase_lite_p2p/couchbase_lite_p2p/CouchbaseLiteP2pPlugin.kt
package com.example.couchbase_lite_p2p.couchbase_lite_p2p

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.couchbase.lite.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.URI

/** CouchbaseLiteP2pPlugin */
class CouchbaseLiteP2pPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    
    private lateinit var database: Database
    private lateinit var multipeerSyncManager: MultipeerSyncManager
    private var replicator: Replicator? = null
    private val scope = CoroutineScope(Dispatchers.Main)
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "couchbase_lite_p2p")
        channel.setMethodCallHandler(this)
        
        initializeCouchbase()
    }

    private fun initializeCouchbase() {
        // Initialize Couchbase Lite
        CouchbaseLite.init(context)
        
        // Open Database
        val config = DatabaseConfiguration()
        database = Database(AppConfig.DATABASE_NAME, config)
        
        // Initialize Sync Manager
        multipeerSyncManager = MultipeerSyncManager(context, database)
        
        // Listen for DB changes
        setupDatabaseListeners()
        
        // Listen for P2P status changes
        setupP2PListeners()
    }

    private fun startSyncGatewayReplication(urlString: String) {
        try {
            val url = URI(urlString)
            val target = URLEndpoint(url)
            // Try using named arguments to avoid ambiguity
            val config = ReplicatorConfiguration(target)
            // Use fully qualified enum if needed, or rely on default (PUSH_AND_PULL)
            // config.replicatorType = ReplicatorType.PUSH_AND_PULL 
            config.isContinuous = true
            // addCollection might require a config object in some versions
            config.addCollection(database.defaultCollection, null)
            
            replicator = Replicator(config)
            
            replicator?.addChangeListener { change ->
                val status = change.status
                val statusStr = when (status.activityLevel) {
                    ReplicatorActivityLevel.STOPPED -> "Stopped"
                    ReplicatorActivityLevel.OFFLINE -> "Offline"
                    ReplicatorActivityLevel.CONNECTING -> "Connecting"
                    ReplicatorActivityLevel.IDLE -> "Idle"
                    ReplicatorActivityLevel.BUSY -> "Busy"
                }
                
                val errorStr = status.error?.message
                
                val statusMap = mapOf(
                    "status" to statusStr,
                    "error" to errorStr
                )
                
                mainHandler.post {
                    channel.invokeMethod("onSyncGatewayStatusChanged", statusMap)
                }
            }
            
            replicator?.start()
        } catch (e: Exception) {
            println("SyncGateway: Failed to start replication: ${e.message}")
            mainHandler.post {
                channel.invokeMethod("onSyncGatewayStatusChanged", mapOf(
                    "status" to "Error",
                    "error" to e.message
                ))
            }
        }
    }

    private fun setupDatabaseListeners() {
        val collection = database.defaultCollection
        collection.addDocumentChangeListener("counter") { change ->
            val doc = collection.getDocument(change.documentID)
            if (doc != null) {
                val count = doc.getInt("count")
                mainHandler.post {
                    channel.invokeMethod("onCountChanged", count)
                }
            }
        }
    }

    private fun setupP2PListeners() {
        scope.launch {
            multipeerSyncManager.syncState.collect { state ->
                val statusMap = mutableMapOf(
                    "isRunning" to state.isRunning,
                    "myPeerID" to (state.myPeerID ?: ""),
                    "connectedPeers" to state.connectedPeers.size,
                    "status" to state.syncStatus
                )
                if (state.error != null) {
                    statusMap["error"] = state.error
                }
                mainHandler.post {
                    channel.invokeMethod("onP2PStatusChanged", statusMap)
                }
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "increment" -> {
                scope.launch {
                    try {
                        val newCount = incrementCounter()
                        result.success(newCount)
                    } catch (e: Exception) {
                        result.error("DB_ERROR", e.message, null)
                    }
                }
            }
            "getCount" -> {
                scope.launch {
                    try {
                        val count = getCounter()
                        result.success(count)
                    } catch (e: Exception) {
                        result.error("DB_ERROR", e.message, null)
                    }
                }
            }
            "startP2P" -> {
                multipeerSyncManager.start()
                result.success(true)
            }
            "getP2PStatus" -> {
                val state = multipeerSyncManager.syncState.value
                val statusMap = mutableMapOf(
                    "isRunning" to state.isRunning,
                    "myPeerID" to (state.myPeerID ?: ""),
                    "connectedPeers" to state.connectedPeers.size,
                    "status" to state.syncStatus
                )
                if (state.error != null) {
                    statusMap["error"] = state.error
                }
                result.success(statusMap)
            }
            "startSyncGatewayReplication" -> {
                val url = call.argument<String>("url")
                if (url != null) {
                    startSyncGatewayReplication(url)
                    result.success(true)
                } else {
                    result.error("INVALID_ARGS", "URL is required", null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }
    
    private suspend fun incrementCounter(): Int = withContext(Dispatchers.IO) {
        val collection = database.defaultCollection
        val doc = collection.getDocument("counter")?.toMutable() ?: MutableDocument("counter")
        val count = doc.getInt("count")
        val newCount = count + 1
        doc.setInt("count", newCount)
        collection.save(doc)
        return@withContext newCount
    }
    
    private suspend fun getCounter(): Int = withContext(Dispatchers.IO) {
        val collection = database.defaultCollection
        val doc = collection.getDocument("counter")
        return@withContext doc?.getInt("count") ?: 0
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        multipeerSyncManager.stop()
        replicator?.stop()
        try {
            database.close()
        } catch (e: Exception) {
            // Ignore
        }
    }
}
